### ISSUE TITLE
Testability of playoutDelayHint

### ISSUE THREAD

**Issue by henbos** [w3c/webrtc-extensions#12]
While the hint may be ignored in various circumstances, there should be enough specificity here to write some basic web platform tests.

Such as measuring jitterBufferDelay / jitterBufferEmittedCount and ensuring the value is &gt;= the hint is achieved, etc. Let's write WPTs and, if anything needs to be clarified or reworded in the spec to ensure the test is interoperable with the spec definition, we do that too.

@kuddai

**Comment by henbos** [comment]
Related to #10, but we can wait that within reasonable testing timeouts we should eventually see that the hint is achieved under ideal circumstances.

**Comment by fippo** [comment]
single-page tests are actually ideal circumstances - we can only test those however (at least in WPT).
Not sure if you can use the WebRTCFakeNetwork field trials in chrome tests.

Made http://fippo.github.io/webrtc//src/content/peerconnection/playoutDelayHint/index.html for playing with this (one needs to manually set in the console, e.g. `pc2.getReceivers()[1].jitterBufferDelayHint = 2.2. Looks like the stats are not on the right object yet because you axed track stats? Otherwise this should be somewhat testable.

**Comment by henbos** [comment]
> Looks like the stats are not on the right object yet because you axed track stats?

A bunch of stuff recently moved in the spec from "track" to "outbound-rtp" (to support simulcast) and "inbound-rtp" (for consistency). "track" should be removed / return duplicate information for backwards-compatiblity reasons, but we've not updated out in/outbound-rtps yet.

**Comment by henbos** [comment]
Cool, here's the https version of the link or else gUM fails ;) https://fippo.github.io/webrtc//src/content/peerconnection/playoutDelayHint/index.html

**Comment by henbos** [comment]
In PR #15 we clarified that this can be measured with jitterBufferDelay/jitterBufferEmittedCount, so we're almost ready to close this issue.

But @jan-ivar [asked](https://github.com/henbos/webrtc-extensions/pull/15#discussion_r348249295) "Can we put any limits on how long implemenations are allowed to take here, to make this testable?" and @kuddai replied (see link) that it is hard to say and depends on network conditions, speech type, etc. Chrome's implementation looks at the wave lengths and adjusts samples in a way as to minimize the perceived effect.

I think we just want to get across that a) it may take several seconds for this (so that an implementation does not cut it off and cause stutter), and b) it should not take several minutes. Chrome may take around 20 s, but this area is still very much being experimented on, so we're reluctant to say something very exact.

May I suggest that we say "it may take several seconds to accelerate or decelerate, but it MUST not take more than a minute"? And then we close this issue.

**Comment by henbos** [comment]
(#11 is also related to testability, but the upper limit on buffer size can be discussed separately from lower limit in acceleration/deceleration speed.)

**Comment by jan-ivar** [comment]
@fippo's tests above are 404ing on me. But here's a basic fiddle I wrote https://jsfiddle.net/jib1/q3a74y6p/

When I run it and you

**Comment by jan-ivar** [comment]
[@fippo](https://github.com/fippo)'s tests are 404 for me. But here's a basic fiddle I wrote https://jsfiddle.net/jib1/hgx4otpL/ — it has a checkbox controlling null/non-null and a slider for value when non-null

We're trying to reverse-engineer Chrome's behavior of `playoutDelayHint` [SIC] since the spec is a bit vague.

When I run it in Chrome I see Chrome defaulting to 22-27 ms jitter buffer delay, and seems somewhat stable at times, but I also see it sometimes grows on its own?

Because of that I'm going to document what I tried, without claiming reproducibility. I tried a couple of things:
- If I set value to 0.5 and check the box then I saw the stat grow to 50 over 10-20 seconds, great!
- If I then uncheck the box, it drops to 27 ms in about 2 seconds, also great
- Starting over, if I drag the slider to 0 and then check the box, it doesn't seem to react and stays at 22-27ms
  - I suppose the spec is flexible enough with its "minimum allowed delay and a maximum allowed delay" to say that the libwebrtc implementation is within bounds of the spec here.

Does this match your experience?

> Such as measuring jitterBufferDelay / jitterBufferEmittedCount and ensuring the value is >= the hint is achieved, etc. Let's write WPTs and, if anything needs to be clarified or reworded in the spec to ensure the test is interoperable with the spec definition, we do that too.

I'm wondering, since jitter buffer stats is JS-observable through getStats, if it would be better to define the value to be == instead of >=.  I'll open a new issue on that.

**Comment by fippo** [comment]
https://github.com/fippo/webrtc/commit/6f77f10f8c8ebd02389f202b9d1c1bffbae2db98 has the old commit.

FWIW, reverse engineering this is likely to fail since the behavior is complex.

**Comment by henbos** [comment]
To me it seems like the actual delay from just looking at the video matches very quickly with the delay that I specified, but quite often the "jitter buffer delay" value remains much lower than this delay, even though it is quite clear that the delay is greater from looking at the video?

**Comment by henbos** [comment]
E.g. video delay set to 2,5 seconds and the delay is clearly 2,5 seconds from looking at the video freed, but jitter buffer delay says ~240 ms.

**Comment by jan-ivar** [comment]
@henbos does the value grow if you wait? It takes many seconds for it to catch up, presumably to avoid freeze frames, but is quick to drop. The spec appears to allow this (it's a target value separate from actual).

**Comment by jan-ivar** [comment]
> E.g. video delay set to 2,5 seconds and the delay is clearly 2,5 seconds from looking at the video feed, but jitter buffer delay says ~240 ms.

Hmm, yeah I see that too. What is the purpose of it working like that?

I naively thought the reason for the slowness correction of the jitter buffer depth was to avoid freeze frames, but if I set the delay to 3 seconds, and then check the ☑ the video freezes for 3 seconds, which defeats that rationale.

**Comment by jan-ivar** [comment]
I'm hoping the reported stat is incorrect (or my fiddle has a bug?), because delaying playout beyond what the jitter buffer demands seems pointless, correct?

**Comment by henbos** [comment]
It appears to slowly grow over time, but maybe one millisecond per second on average, so I might have to wait for 15 minutes in real time before the jitter buffer delay has increased by a second... I didn't test that long.

If I reduce the playoutDelayHint then the jitter buffer delay quickly matches the playoutDelayHint.

Agree this seems buggy, I would not expect it to affect the jitter buffer when the hint is decreasing but not when increasing. But I don't know what our implementation does

**Comment by henbos** [comment]
I think it makes sense for it taking a few or even several seconds for the delay to reach our hint to avoid freezes by reducing playout speed, but speed at which the jitter buffer increases is clearly not the speed of the playout delay. It makes you wonder: is the metric broken, or is there additional buffer happening that is not accounted for in the jitter buffer?

**Comment by jan-ivar** [comment]
> I would not expect it to affect the jitter buffer when the hint is decreasing but not when increasing.

Well, _if_ a one second change in jitter buffer depth affected playback, then a one second skip ahead (avoiding audio clicks) might be tolerable in some use cases, whereas a second of frozen silence maybe less so? (speculation)

> It makes you wonder: is the metric broken, or is there additional buffer happening that is not accounted for in the jitter buffer?

This is my question as well. I think we have a spec problem because the implementation appears to pass when it shouldn't:
1. [playoutDelay](https://w3c.github.io/webrtc-extensions/#dom-rtcrtpreceiver-playoutdelay) says explicitly _"The [user agent](https://w3c.github.io/webrtc-extensions/#dfn-user-agent) SHOULD NOT playout audio or video that is received unless this amount of time has passed in seconds"_
    - ✅ PASS (talks only about playout)

3. In parallel: _"If the given delay value is ... above [maximum allowed delay](https://w3c.github.io/webrtc-extensions/#dfn-maximum-allowed-delay) then the value used MUST be clamped"_
    - ✅ PASS (talks about the _"value used"_ below for internal buffering) 

6. _"Modifying the delay of the underlying system SHOULD affect the internal audio or video buffering gradually in order not to hurt user experience. Audio samples or video frames SHOULD be accelerated or decelerated before playout, similarly to how it is done for[ audio/video synchronization](https://www.rfc-editor.org/rfc/rfc5888#section-7) or in response to congestion control."_
    - ✅ PASS (talks about internal buffering "before" playout)

I think the problem is that from a pedantic read there are two independently observable effects: jitter and delay.
- Statement 1 talks only about playout.
- Statement 2 and 3 appear to ultimately only talk about internal (jitter) buffering

If these effects are allowed to be controlled independently, then the implementation passes.

I think this is wrong.

### What is the desired effect?

IMHO the API should only chase one goal effect: jitter-free media. Delay is an undesirable side-effect of that. This means:
1. Statement 1 and 2 seem contradictory (is effect clamped or not?)
2. Statement 1 and 3 seem contradictory (is effect gradual or not?)
3. Any non-jitter-buffer buffering defeats this goal (a wasted opportunity to handle NACKs)

`playoutDelay` may have a terminology problem, e.g. like if my car had a drag-ometer. This is why I opened #156.

**Comment by jan-ivar** [comment]
> Any non-jitter-buffer buffering defeats this goal (a wasted opportunity to handle NACKs)

We're speculating AV sync may be one reason for libwebrtc doing this.

**Comment by jan-ivar** [comment]
Here's an updated fiddle with stats for both audio and video https://jsfiddle.net/jib1/0L6duga2/

**Comment by henbos** [comment]
If we can get alignment on this being a control surface for target jitter buffer delay rather than "playout" then the the API would be more testable.

If the name is also wrong (playoutDelayHint vs playoutDelay), what do you think about changing the name to desiredJitterBufferDelay or something like that? (Avoiding the word "target" because the internal target is already exposed as [jitterBufferTargetDelay](https://w3c.github.io/webrtc-stats/#dom-rtcinboundrtpstreamstats-jitterbuffertargetdelay) and this value is not necessarily the same as the one the app asked for)

**Comment by henbos** [comment]
Some discussion later and I think jitterBufferDelay simply does not include the entire delay in the case of video and that we should just update the metric

**Comment by henbos** [comment]
jitterBufferDepth sgtm if this becomes observable in jitterBufferDelay, which it probably will be after a bugfix

**Comment by henbos** [comment]
https://crbug.com/webrtc/15085 has been filed and assigned, reflecting our intentions to align
